### ISSUE TITLE
When rendering coloured Terracotta, not all colours are rendered properly

### ISSUE THREAD

**Issue by karpana** [mjungnickel18/papyruscs#55]
As noted above, I have a world where I have extensive use of Terracotta for building.
I can say with surety, that Red Terracotta, Black and Brown, all appear to render in "white".

**Comment by karpana** [comment]
What I don't know is whether this issue has to do with the hiegh of the buildings?  Or whether they are just rendering as white.

**Comment by bennydiamond** [comment]
I strongly believe issue is coming from [CoreFNBT](https://www.nuget.org/packages/CoreFNBT/) library used in this project.

It has not been updated to support multiple colors of terracotta (previously named stained_hardened_clay) and thus, will always parse block data to contain an "Item Data" value of 0, which correspond to white terracotta according to [this table.](https://minecraftitemids.com/types/clay).

Whatever the terracotta block color, the block is always identified as _minecraft:stained_hardened_clay:0_ and thus displays as white terracotta.

**Comment by karpana** [comment]
I guess the question I have then, is how does this get fixed?

**Comment by bennydiamond** [comment]
It would seem @aaronaxvig has done [some work to support mulitple NBT file versions](https://github.com/aaronaxvig/fNbt-Core/commit/7fb8975f38646d34329927dac53868629dad1312). 

Maybe that's the key? But is it working?

**Comment by vtbassmatt** [comment]
Concrete is the same - always renders as white.

**Comment by karpana** [comment]
@bennydiamond 
With you having made some changes (I see the merge from a few days ago).
Is there a way for me to pull these in to my copy, given that I"m using the 0.5.0 pre-compiled binaries on my linux server.

**Comment by bennydiamond** [comment]
@karpana checkout master branch and build following instructions in readme. 

@mjungnickel18 recently merged my PRs, a new release should be coming in soon but it would be a nice contribution if you could confirm which issues are resolved by the new changes beforehand.

**Comment by karpana** [comment]
@bennydiamond 

Gave this a shot.
the installtion of DotNet seemed to go okay on my Ubuntu machine (following Microsoft's instructions).
the compilation of PapyrusCs seemed to go okay following the read me.

the execution ... outright fell apart....



`Opening world...
Generating a list of all chunk keys in the database.
This could take a few minutes
Total Chunk count 113508

The total dimensions of the map are
  X: -655 to 1135
  Z: -654 to 1003

Tilesize is 512x512
Reading terrain_texture.json...

The maximum diameter of the map is 1791
For 2 chunks per tile, new max diameter is 896
To generate the zoom levels, we expand the diameter to 1024
This results in 11 zoom levels
Found 0 subchunks which are already rendered
Grouping subchunks... 113508
Grouped by 2 to 32446 tiles
Average of 3.5 chunks per tile
519 of 113508 Chunks render @ 114.8 cps
Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 19 Db 0

Queue Stat: GetData 0 Render 0 Save 18 Db 0

Queue Stat: GetData 0 Render 0 Save 18 Db 0

Queue Stat: GetData 0 Render 0 Save 18 Db 0

Queue Stat: GetData 0 Render 0 Save 18 Db 0

Queue Stat: GetData 0 Render 0 Save 17 Db 0

Queue Stat: GetData 0 Render 0 Save 17 Db 0

Queue Stat: GetData 0 Render 0 Save 17 Db 0

Queue Stat: GetData 0 Render 0 Save 17 Db 0

Queue Stat: GetData 0 Render 0 Save 17 Db 0

Queue Stat: GetData 0 Render 0 Save 17 Db 0
Post complete
648 of 113508 Chunks render @ 74.5 cps     Unhandled exception. System.AggregateException: One or more errors occurred. (One or more errors occurred. (One or more errors occurred. (One or more errors occurred. (One or more errors occurred. (IO error: /home/minecraft/git/papyruscs/PapyrusCs/bin/Debug/netcoreapp3.1/linux-x64/worlds/Bedrock Level/db/063273.ldb: Invalid argument)))))
 ---> System.AggregateException: One or more errors occurred. (One or more errors occurred. (One or more errors occurred. (One or more errors occurred. (IO error: /home/minecraft/git/papyruscs/PapyrusCs/bin/Debug/netcoreapp3.1/linux-x64/worlds/Bedrock Level/db/063273.ldb: Invalid argument))))
 ---> System.AggregateException: One or more errors occurred. (One or more errors occurred. (One or more errors occurred. (IO error: /home/minecraft/git/papyruscs/PapyrusCs/bin/Debug/netcoreapp3.1/linux-x64/worlds/Bedrock Level/db/063273.ldb: Invalid argument)))
 ---> System.AggregateException: One or more errors occurred. (One or more errors occurred. (IO error: /home/minecraft/git/papyruscs/PapyrusCs/bin/Debug/netcoreapp3.1/linux-x64/worlds/Bedrock Level/db/063273.ldb: Invalid argument))
 ---> System.AggregateException: One or more errors occurred. (IO error: /home/minecraft/git/papyruscs/PapyrusCs/bin/Debug/netcoreapp3.1/linux-x64/worlds/Bedrock Level/db/063273.ldb: Invalid argument)
 ---> System.ApplicationException: IO error: /home/minecraft/git/papyruscs/PapyrusCs/bin/Debug/netcoreapp3.1/linux-x64/worlds/Bedrock Level/db/063273.ldb: Invalid argument
   at leveldb_sharp_std.Native.CheckError(String error) in /home/minecraft/git/papyruscs/leveldb-sharp-std/Native.cs:line 52
   at leveldb_sharp_std.Native.leveldb_get(IntPtr db, IntPtr readOptions, Byte[] key, UIntPtr& valueLength) in /home/minecraft/git/papyruscs/leveldb-sharp-std/Native.cs:line 216
   at leveldb_sharp_std.DB.Get(ReadOptions options, Byte[] key, UIntPtr& length) in /home/minecraft/git/papyruscs/leveldb-sharp-std/DB.cs:line 193
   at leveldb_sharp_std.DB.Get(Byte[] key, UIntPtr& length) in /home/minecraft/git/papyruscs/leveldb-sharp-std/DB.cs:line 198
   at Maploader.World.World.GetChunkData(GroupedChunkSubKeys groupedChunkSubKeys) in /home/minecraft/git/papyruscs/Maploader/World/World.cs:line 515
   at PapyrusAlgorithms.Strategies.Dataflow.GetDataBlock.<>c__DisplayClass3_0.<.ctor>b__0(IEnumerable`1 groupedChunkSubKeys) in /home/minecraft/git/papyruscs/PapyrusAlgorithms/Strategies/Dataflow/GetDataBlock.cs:line 24
   at System.Threading.Tasks.Dataflow.TransformManyBlock`2.ProcessMessage(Func`2 transformFunction, KeyValuePair`2 messageWithId)
   at System.Threading.Tasks.Dataflow.Internal.TargetCore`1.ProcessMessagesLoopCore()
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout)
   at PapyrusAlgorithms.Strategies.Dataflow.DataFlowStrategy`1.RenderInitialLevel() in /home/minecraft/git/papyruscs/PapyrusAlgorithms/Strategies/Dataflow/DataFlowStrategy.cs:line 193
   at PapyrusCs.Program.RunMapCommand(Options options) in /home/minecraft/git/papyruscs/PapyrusCs/Program.cs:line 234
   at PapyrusCs.Program.<>c.<Main>b__4_0(Options opts) in /home/minecraft/git/papyruscs/PapyrusCs/Program.cs:line 45
   at CommandLine.ParserResultExtensions.MapResult[T1,T2,T3,TResult](ParserResult`1 result, Func`2 parsedFunc1, Func`2 parsedFunc2, Func`2 parsedFunc3, Func`2 notParsedFunc)
   at PapyrusCs.Program.Main(String[] args) in /home/minecraft/git/papyruscs/PapyrusCs/Program.cs:line 40
Aborted (core dumped)
`

**Comment by bennydiamond** [comment]
@karpana Could you upload a copy of your world on a file hoster of your choice?

**Comment by karpana** [comment]
the 063723.ldb file *does* exist in my folder.
interestingly, it's the only file that has zero bytes (which seems might unusual).

A quick test with the pre-compiled 0.50 provides the same result.  
hmmm...

And a quick load of my world shows definite corruption (4 or 5 totally missing, and now regenerated chunks).
I'm going to load a backup and try all this again.

**Comment by karpana** [comment]
Found a "clean" backup (had to go back a few days).
And now have a "clean" run of 0.5.0.

Thank you to @bennydiamond for the tip.
and gave it a whirl with the dev version that I compiled.

This time it ran clean from front to back (like the 0.5.0) version.
I"m attaching a screen shot that shows the badlands area of my world where I've been doing most of my playing. 
The differences are most noticeable North-East of the "Scar" in the top-middle of my screen shot.

It looks like the PullRequest that was completed solves the issues.
Thank you @barrett777 & @aaronaxvig for the fix.
![PapyrusCs_cc32985](https://user-images.githubusercontent.com/15716173/87234831-8f9c8180-c39a-11ea-964b-94ef4a3962c6.jpg)
